File: native/ios/Sources/Walk.swift
import AVFoundation
import Foundation
import Speech
import SwiftUI

/// The conversation that keeps going when this app is not on screen.
///
/// The console's walk-along strip carries a conversation across a screen
/// change and stops dead when the browser puts the page away. That is not a
/// shortcoming of the strip — a backgrounded web page has its recogniser
/// ended for it — and the strip says so rather than pretending.
///
///     asked     can the conversation survive a screen change
///     mattered  can it survive leaving the application
///
/// On a phone the answer is yes, and iOS has a well-worn path for it: the
/// `audio` background mode, an `AVAudioSession` that stays active, and the
/// orange indicator the system puts in the status bar for as long as the
/// microphone is open. That indicator is the point rather than a side effect.
/// It is what makes this *the conversation you took with you* instead of *an
/// app recording you after you left it*, and the two are the same code with
/// different honesty. Android buys the same honesty with a notification it
/// cannot dismiss; here the system draws it, which is better, because a
/// person learns one indicator for every app rather than one per app.
///
/// ## Why the session is configured the way it is
///
/// * `.playAndRecord` — the reply plays while the microphone is open, which
///   is what lets somebody interrupt. A conversation you cannot interrupt is
///   a broadcast.
/// * `.spokenAudio` — tells the system this is speech rather than music, so
///   it is handled like a podcast or a navigation voice.
/// * `.mixWithOthers` and `.duckOthers` — somebody walking about with this
///   open may well have music on. Stopping their music dead would be this
///   app deciding it is the only thing that matters; ducking says its piece
///   and gives the room back.
/// * `.allowBluetooth` / `.allowBluetoothA2DP` — the earbud already in their
///   ear. The console asks for the connected microphone by name for the same
///   reason, after a field report had somebody pulling an earbud out to hear
///   their own guardian.
///
/// ## Written without a compiler
///
/// There is no Swift toolchain in the environment this was written in, and
/// no Android SDK either — the proxy refuses `dl.google.com` outright. So
/// the guard beside this reads the declarations rather than the behaviour:
/// the background mode, the usage strings, the session category, the turn
/// scoping, and that stopping actually deactivates the session. Those are
/// the parts whose absence is a microphone with no indicator, or an
/// indicator that never goes out. The loop has been reasoned about and not
/// run, and it wants a person with a phone before anybody calls it working.
@MainActor
final class Walking: ObservableObject {

    static let shared = Walking()

    /// Whether a walk is in progress, for the screen that offers the button.
    @Published private(set) var underway = false
    /// The last thing heard and the last thing said, so somebody coming back
    /// to the app sees where the conversation got to rather than the screen
    /// exactly as they left it.
    @Published private(set) var heard = ""
    @Published private(set) var said = ""
    /// Why it stopped, when it stopped for a reason. Empty when somebody
    /// ended it on purpose — they know.
    @Published private(set) var trouble = ""
    /// Bumped every time a walk begins, so the shell can land the person on
    /// the front page. The point of taking a conversation with you is going
    /// somewhere, and the screen you were on is the one place you have
    /// finished with — leaving somebody on the coach screen with the strip
    /// lit means the first thing they do is find their way out of it.
    ///
    ///     asked     did the conversation survive
    ///     mattered  can they now go anywhere
    ///
    /// A counter rather than a flag: a second walk started from the front
    /// page must still land there, and a boolean already true would say
    /// nothing happened.
    @Published private(set) var landings = 0

    /// True when the last turn was answered by the offline stack rather than
    /// by a model. Not a failure — a deployment with no model key still
    /// coaches, from stored knowledge — but not the model somebody picked
    /// either, and out here there is no screen to notice it on.
    @Published private(set) var offline = false

    private let engine = AVAudioEngine()
    private let speaker = AVSpeechSynthesizer()
    private var recogniser: SFSpeechRecognizer?
    private var request: SFSpeechAudioBufferRecognitionRequest?
    private var task: SFSpeechRecognitionTask?
    private var quiet: Timer?

    private var uid = ""
    private var token = ""
    private var area = "general"
    private var lang = "en"

    /// Every opening of the ear carries a number, and a late callback from a
    /// superseded one is ignored. The console learned this the hard way: one
    /// shared flag meant a stale error closed the ear that had just replaced
    /// it, and the microphone died a fifth of a second after it opened.
    private var turn = 0
    private var wants = false

    /// How long a turn waits on quiet before it is taken. The console's own
    /// number, from the same field report: five seconds was "still a long
    /// delay while waiting for a response — drop it to 2.5".
    private let quietSeconds = 2.5

    private init() {}

    // MARK: - Starting and stopping

    /// Take the conversation with you. Called from a button, never from an
    /// appear handler: an ear that outlives its screen without a press is
    /// the headless microphone every teardown in this shell exists to
    /// prevent.
    func start(uid: String, token: String, area: String, lang: String) {
        guard !underway else { return }
        self.uid = uid
        self.token = token
        // `general` is the front door's own area and a fallback here, never
        // a silent replacement for one somebody picked.
        self.area = area.isEmpty ? "general" : area
        self.lang = lang
        trouble = ""
        offline = false

        // Both permissions, and the honest sentence for each refusal. A
        // screen that reports "the microphone stopped" for a permission
        // never granted is the silence-and-deafness confusion this estate
        // keeps finding.
        SFSpeechRecognizer.requestAuthorization { [weak self] speech in
            Task { @MainActor in
                guard let self else { return }
                guard speech == .authorized else {
                    self.trouble = L10n.t("walk.trouble.speech", self.lang)
                    return
                }
                AVAudioSession.sharedInstance().requestRecordPermission { mic in
                    Task { @MainActor in
                        guard mic else {
                            self.trouble =
                                L10n.t("walk.trouble.permission", self.lang)
                            return
                        }
                        self.begin()
                    }
                }
            }
        }
    }

    /// Leave: nothing in flight answers, nothing re-opens, the session goes
    /// back to the system. Deactivating matters as much as activating did —
    /// a session left active is the orange indicator still lit over an app
    /// that has stopped listening, which teaches people the indicator lies.
    func stop() {
        close(reason: "")
    }

    private func begin() {
        guard let recognizer = SFSpeechRecognizer(
            locale: Locale(identifier: lang)) ?? SFSpeechRecognizer(),
              recognizer.isAvailable else {
            trouble = L10n.t("walk.trouble.norecogniser", lang)
            return
        }
        recogniser = recognizer
        do {
            let session = AVAudioSession.sharedInstance()
            try session.setCategory(
                .playAndRecord, mode: .spokenAudio,
                options: [.mixWithOthers, .duckOthers,
                          .allowBluetooth, .allowBluetoothA2DP,
                          .defaultToSpeaker])
            try session.setActive(true, options: [])
        } catch {
            trouble = L10n.t("walk.trouble.session", lang)
            return
        }
        wants = true
        underway = true
        landings += 1
        hear()
    }

    // MARK: - One turn

    private func hear() {
        guard wants, let recognizer = recogniser else { return }
        turn += 1
        let mine = turn
        func live() -> Bool { mine == turn && wants }

        let req = SFSpeechAudioBufferRecognitionRequest()
        req.shouldReportPartialResults = true
        // On-device where the phone offers it. Not an optimisation: a
        // recogniser that needs Apple's servers stops working in the exact
        // place this feature is for — a person out of the app and out of
        // signal — and it also keeps the audio on the phone, which is the
        // better default for a health product whether or not anybody asked.
        if recognizer.supportsOnDeviceRecognition {
            req.requiresOnDeviceRecognition = true
        }
        request = req

        let input = engine.inputNode
        let format = input.outputFormat(forBus: 0)
        input.removeTap(onBus: 0)
        input.installTap(onBus: 0, bufferSize: 1024, format: format) { buffer, _ in
            req.append(buffer)
        }
        engine.prepare()
        do {
            try engine.start()
        } catch {
            close(reason: L10n.t("walk.trouble.session", lang))
            return
        }

        task = recognizer.recognitionTask(with: req) { [weak self] result, error in
            Task { @MainActor in
                guard let self, live() else { return }
                if let result {
                    let text = result.bestTranscription.formattedString
                    self.heard = text
                    // Quiet ends a turn, the way the console's does: the tap
                    // keeps feeding an open request otherwise, and a person
                    // who has finished a sentence is left waiting on a
                    // recogniser that does not know they stopped.
                    self.restartQuietTimer(mine: mine)
                    if result.isFinal, !text.isEmpty {
                        self.take(mine: mine, message: text)
                    }
                    return
                }
                if error != nil {
                    // Nothing heard is not a failure in a standing
                    // conversation — the microphone simply opens again.
                    // Everything else stops and says which failure it was,
                    // because a refusal reported as quiet is a loop that
                    // reopens forever with nothing to hear.
                    if self.heard.isEmpty {
                        self.hear()
                    } else {
                        self.close(
                            reason: L10n.t("walk.trouble.stopped", self.lang))
                    }
                }
            }
        }
    }

    private func restartQuietTimer(mine: Int) {
        quiet?.invalidate()
        quiet = Timer.scheduledTimer(withTimeInterval: quietSeconds,
                                     repeats: false) { [weak self] _ in
            Task { @MainActor in
                guard let self, mine == self.turn, self.wants else { return }
                // End the audio rather than the task: the recogniser then
                // delivers its final transcription, which is the turn.
                self.request?.endAudio()
            }
        }
    }

    private func take(mine: Int, message: String) {
        endListening()
        Task {
            let reply = try? await ApiClient.shared.coach(
                uid: uid, token: token, area: area, message: message)
            await MainActor.run {
                guard mine == self.turn, self.wants else { return }
                let text = reply?.content ?? ""
                // Who actually answered, not who was picked. That
                // distinction is the whole reason the field exists.
                self.offline = reply?.provenance?.generated_by == "stub"
                if text.isEmpty {
                    self.said = L10n.t("walk.lost", self.lang)
                } else {
                    self.said = text
                    let utterance = AVSpeechUtterance(string: text)
                    utterance.voice =
                        AVSpeechSynthesisVoice(language: self.lang)
                    self.speaker.speak(utterance)
                }
                self.heard = ""
                // The next turn opens with the voice rather than after it: a
                // person may interrupt, and `.playAndRecord` is what makes
                // that possible in the first place.
                self.hear()
            }
        }
    }

    // MARK: - Tearing down

    private func endListening() {
        quiet?.invalidate()
        quiet = nil
        engine.inputNode.removeTap(onBus: 0)
        if engine.isRunning { engine.stop() }
        request?.endAudio()
        request = nil
        task?.cancel()
        task = nil
    }

    private func close(reason: String) {
        wants = false
        turn += 1
        endListening()
        speaker.stopSpeaking(at: .immediate)
        // Give the session back. A session left active keeps the orange
        // indicator lit over an app that is no longer listening, and an
        // indicator that lies is worse than none.
        try? AVAudioSession.sharedInstance().setActive(
            false, options: [.notifyOthersOnDeactivation])
        underway = false
        offline = false
        trouble = reason
    }
}

File: native/windows/Walk.cs
using System;
using System.Threading.Tasks;
using Windows.Media.SpeechRecognition;
using Windows.Media.SpeechSynthesis;
using Microsoft.UI.Xaml.Controls;

namespace JimGuardian;

/// <summary>
/// The conversation that keeps going while this window is not the one you
/// are looking at.
///
/// <para>
/// Windows is the odd one out of the three shells, and saying why matters
/// more than the code does. A phone suspends an app the moment it leaves the
/// screen, so Android needs a foreground service and iOS needs a background
/// audio mode — both of them permission to keep running at all. A minimised
/// desktop window is not suspended. This app is unpackaged (see
/// <c>WindowsPackageType=None</c> in the csproj), so it does not even take
/// the packaged app lifecycle: minimise it and it keeps running, exactly as
/// it did before.
/// </para>
///
/// <para>
///     asked     can the conversation survive a screen change<br/>
///     mattered  can it survive leaving the application
/// </para>
///
/// <para>
/// So there was never an operating system to satisfy here. What was missing
/// was a voice loop: this shell had none at all, and a conversation that
/// cannot hear you does not become carryable by the window staying open. That
/// is what this is.
/// </para>
///
/// <para>
/// The honesty this buys elsewhere with a notification or an orange dot is
/// bought here by Windows' own microphone indicator in the system tray, and
/// by the control that started the walk saying <em>End</em> for as long as it
/// runs. There is no supported way for a desktop app to open a microphone
/// without that tray indicator, which is the right arrangement.
/// </para>
///
/// <para>
/// <b>Written without a compiler.</b> There is no .NET SDK in the environment
/// this was written in. The guard beside it reads what can be read from the
/// source: that a loop exists, that it is scoped by turn, that quiet reopens
/// and a refusal does not, and that stopping actually disposes the recogniser
/// rather than leaving the tray indicator lit over an app that has stopped
/// listening. The loop itself has been reasoned about and not run.
/// </para>
/// </summary>
public static class Walking
{
    /// <summary>Raised whenever anything below changes, so a page can redraw
    /// without polling.</summary>
    public static event Action? Changed;

    public static bool Underway { get; private set; }
    public static string Heard { get; private set; } = "";
    public static string Said { get; private set; } = "";
    /// <summary>Why it stopped, when it stopped for a reason. Empty when
    /// somebody ended it on purpose — they know.</summary>
    public static string Trouble { get; private set; } = "";
    /// <summary>True when the last turn was answered by the offline stack
    /// rather than by a model. Not a failure — a deployment with no model key
    /// still coaches, from stored knowledge — but not the model somebody
    /// picked either.</summary>
    public static bool Offline { get; private set; }
    /// <summary>Bumped when a walk begins, so the shell can land the person on
    /// the front page. A counter rather than a flag: a second walk started
    /// from the front page must still land there.</summary>
    public static int Landings { get; private set; }

    private static SpeechRecognizer? _recogniser;
    private static SpeechSynthesizer? _speaker;
    private static MediaElement? _voice;
    private static string _uid = "", _token = "", _area = "general", _lang = "en";

    /// <summary>Every opening of the ear carries a number, and a late callback
    /// from a superseded one is ignored. The console learned this the hard
    /// way: one shared flag meant a stale error closed the ear that had just
    /// replaced it.</summary>
    private static int _turn;
    private static bool _wants;

    /// <summary>Take the conversation with you. Called from a press, never
    /// from a page loading: an ear that outlives its screen without a press is
    /// a microphone nobody asked for.</summary>
    public static async Task Start(string uid, string token, string area,
                                   string lang, MediaElement voice)
    {
        if (Underway) return;
        _uid = uid; _token = token;
        // `general` is the front door's own area, and a fallback here rather
        // than a silent replacement for one somebody picked.
        _area = string.IsNullOrWhiteSpace(area) ? "general" : area;
        _lang = lang;
        _voice = voice;
        Trouble = ""; Offline = false;

        try
        {
            _recogniser = new SpeechRecognizer();
            await _recogniser.CompileConstraintsAsync();
            _speaker = new SpeechSynthesizer();
        }
        catch (Exception)
        {
            // A refused microphone and a machine with no recogniser both land
            // here, and both get a sentence rather than silence: somebody who
            // has just pressed a button and heard nothing cannot tell the two
            // apart, and only one of them is theirs to fix.
            Trouble = L10n.T("walk.trouble.norecogniser", _lang);
            Changed?.Invoke();
            return;
        }

        _wants = true;
        Underway = true;
        Landings += 1;
        Changed?.Invoke();
        await Hear();
    }

    /// <summary>Leave: nothing in flight answers, nothing reopens, and the
    /// recogniser is disposed so the tray indicator goes out. An indicator
    /// left lit over an app that has stopped listening teaches people the
    /// indicator lies.</summary>
    public static void Stop() => Close("");

    private static async Task Hear()
    {
        if (!_wants || _recogniser is null) return;
        var mine = ++_turn;
        SpeechRecognitionResult result;
        try
        {
            result = await _recogniser.RecognizeAsync();
        }
        catch (Exception)
        {
            if (mine != _turn || !_wants) return;
            Close(L10n.T("walk.trouble.stopped", _lang));
            return;
        }
        if (mine != _turn || !_wants) return;

        // Quiet is not a failure in a standing conversation — the microphone
        // simply opens again. Everything else stops and says which failure it
        // was, because a refusal reported as quiet is a loop that reopens
        // forever with nothing to hear.
        if (result.Status != SpeechRecognitionResultStatus.Success)
        {
            if (result.Status == SpeechRecognitionResultStatus.TimeoutExceeded)
            {
                await Hear();
                return;
            }
            Close(L10n.T("walk.trouble.stopped", _lang));
            return;
        }
        var text = (result.Text ?? "").Trim();
        if (text.Length == 0) { await Hear(); return; }

        Heard = text;
        Changed?.Invoke();
        await Take(mine, text);
    }

    private static async Task Take(int mine, string message)
    {
        string reply = "";
        bool fromStore = false;
        try
        {
            var g = await ApiClient.Shared.Coach(_uid, _token, _area, message);
            reply = g.Content ?? "";
            // Who actually answered, not who was picked. That distinction is
            // the whole reason the field exists.
            fromStore = g.ProvenanceInfo?.GeneratedBy == "stub";
        }
        catch (Exception)
        {
            reply = "";
        }
        if (mine != _turn || !_wants) return;

        Offline = fromStore;
        if (reply.Length == 0)
        {
            Said = L10n.T("walk.lost", _lang);
        }
        else
        {
            Said = reply;
            Heard = "";
            try
            {
                if (_speaker is not null && _voice is not null)
                {
                    var stream = await _speaker.SynthesizeTextToStreamAsync(reply);
                    _voice.SetSource(stream, stream.ContentType);
                    _voice.Play();
                }
            }
            catch (Exception)
            {
                // A voice that will not play is not a reason to end a
                // conversation somebody is having in writing on the screen
                // behind them. The words are already in `Said`.
            }
        }
        Changed?.Invoke();
        // The next turn opens with the voice rather than after it: a person
        // may interrupt, and a conversation that cannot be interrupted is a
        // broadcast.
        await Hear();
    }

    private static void Close(string reason)
    {
        _wants = false;
        _turn += 1;
        try { _voice?.Stop(); } catch (Exception) { }
        _recogniser?.Dispose();
        _recogniser = null;
        _speaker?.Dispose();
        _speaker = null;
        Underway = false;
        Offline = false;
        Trouble = reason;
        Changed?.Invoke();
    }
}
